server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java

```diff
@@ -129,11 +138,29 @@ private void sendUnpromotableRequests(IndexShard indexShard, long generation, bo
             transportService.getLocalNode(),
             TransportUnpromotableShardRefreshAction.NAME,
             unpromotableReplicaRequest,
-            new ActionListenerResponseHandler<>(
-                listener.delegateFailure((l, r) -> l.onResponse(wasForced)),
-                (in) -> ActionResponse.Empty.INSTANCE,
-                ThreadPool.Names.REFRESH
-            )
+            new ActionListenerResponseHandler<>(ActionListener.wrap(r -> listener.onResponse(wasForced), e -> {
+                ShardRouting shardRouting = indexShard.routingEntry();
```

Comment:
I still believe this is failing the wrong shard. We are just getting the routing entry from the local index shard and failing it. I believe that is failing the indexing shard?

I think that we would need to propagate the replicas that fail in the TransportUnpromotableShardRefreshAction response and fail those (it can be more than 1).

If there is an exception level error, we don't know if any of the search shards succeeded and need to fail all of them.

Reply:
Unless I am missing something here?

Since we can have multiple unpromotable shards and some can fail and some can succeed, I definitely think there would need to be some work here around failing multiple shards.

Reply:
BTW - It might be easier to back this into `TransportBroadcastUnpromotableAction`. Essentially make a config on the request that allows you to request that an error response shard is failed. That would remove the need to inject ShardStateAction in here.

Reply:
I've submitted #95049 with the proposed implementation